identity-mdoc/src/main/java/com/android/identity/mdoc/mso/MobileSecurityObjectGenerator.kt

```diff
@@ -0,0 +1,276 @@
+    private var digestEmpty = true
+    private val mAuthorizedNameSpaces: MutableList<String> = ArrayList()
+    private val mAuthorizedDataElements: MutableMap<String, List<String>> = HashMap()
+    private val mKeyInfo: MutableMap<Long, ByteArray> = HashMap()
+    private var mSigned: Timestamp? = null
+    private var mValidFrom: Timestamp? = null
+    private var mValidUntil: Timestamp? = null
+    private var mExpectedUpdate: Timestamp? = null
+
+    init {
+        val allowableDigestAlgorithms = listOf("SHA-256", "SHA-384", "SHA-512")
+        require(allowableDigestAlgorithms.contains(digestAlgorithm)) {
+            "digestAlgorithm must be one of $allowableDigestAlgorithms"
+        }
+        mDigestAlgorithm = digestAlgorithm
+        mDocType = docType
+        mDeviceKey = deviceKey
+        mDigestSize = when (digestAlgorithm) {
+            "SHA-256" -> 32
+            "SHA-384" -> 48
+            "SHA-512" -> 64
+            else -> -1
+        }
+    }
+
+    /**
+     * Populates the `ValueDigests` mapping. This must be called at least once before
+     * generating since `ValueDigests` must be non-empty.
+     *
+     * @param nameSpace The namespace.
+     * @param digestIDs A non-empty mapping between a `DigestID` and a `Digest`.
+     * @return The `MobileSecurityObjectGenerator`.
+     * @exception IllegalArgumentException if the `digestIDs` is empty.
+     */
+    fun addDigestIdsForNamespace(
+        nameSpace: String,
+        digestIDs: Map<Long, ByteArray>
+    ) = apply {
+        require(!digestIDs.isEmpty()) { "digestIDs must not be empty" }
+        digestEmpty = false
+        val valueDigestsInner = mValueDigestsOuter.putMap(nameSpace)
+        for (digestID in digestIDs.keys) {
+            val digest = digestIDs[digestID]
+            require(digest!!.size == mDigestSize) {
+                "digest is unexpected length: expected $mDigestSize, got ${digest.size}"
+            }
+            valueDigestsInner.put(digestID, digest)
+        }
+        valueDigestsInner.end()
+    }
+
+    /**
+     * Populates the `AuthorizedNameSpaces` portion of the `keyAuthorizations`
+     * within `DeviceKeyInfo`. This gives authorizations to full namespaces
+     * included in the `authorizedNameSpaces` array. If authorization is given for a full
+     * namespace, that namespace shall not be included in
+     * [.setDeviceKeyAuthorizedDataElements].
+     *
+     * @param authorizedNameSpaces A list of namespaces which should be given authorization.
+     * @return The `MobileSecurityObjectGenerator`.
+     * @throws IllegalArgumentException if the authorizedNameSpaces does not meet the constraints.
+     */
+    fun setDeviceKeyAuthorizedNameSpaces(
+        authorizedNameSpaces: List<String>
+    ) = apply {
+        val namespaceSet: MutableSet<String> = HashSet()
+        namespaceSet.addAll(mAuthorizedDataElements.keys)
+        namespaceSet.retainAll(authorizedNameSpaces)
```

Comment:
Kodlin DSL 
- making use of `mutableSetOf()` 
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/mutable-set-of.html
- using `apply` 
```
        val namespaceSet = mutableSetOf<String>().apply {
            addAll(mAuthorizedDataElements.keys)
            retainAll(authorizedNameSpaces)
        }
```

Reply:
(I'm adding this to the follow-up PR)